x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java

```diff
@@ -390,22 +408,50 @@ public void setReindexingTaskId(Long reindexingTaskId) {
             this.reindexingTaskId = reindexingTaskId;
         }
 
-        @Nullable
-        public Long getReindexingTaskId() {
-            return reindexingTaskId;
+        public void setReindexingFinished() {
+            isReindexingFinished = true;
         }
 
         public boolean isStopping() {
             return isStopping;
         }
 
+        public ProgressTracker getProgressTracker() {
+            return progressTracker;
+        }
+
         @Override
         protected void onCancelled() {
             stop(getReasonCancelled(), TimeValue.ZERO);
         }
 
+        @Override
+        public void markAsCompleted() {
+            persistProgress(ActionListener.wrap(() -> super.markAsCompleted()));
+        }
+
+        @Override
+        public void markAsFailed(Exception e) {
+            persistProgress(ActionListener.wrap(() -> super.markAsFailed(e)));
```

Comment:
I sort of "ham 'n haw" around whether we should always use an `ActionListener` or if there is ever justification to use `Consumer` function. In this case, it seems to me that we want to prevent `listener.onFailure` from EVERY being called as it could cause the task to stick in limbo. 

Maybe just use a function here?

Reply:
Direct `Runnable` it is! :+1: